clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -339,24 +338,21 @@ private TransactionalRequestResult beginCompletingTransaction(TransactionResult
         if (!newPartitionsInTransaction.isEmpty())
             enqueueRequest(addPartitionsToTransactionHandler());
 
-        // If the error is an INVALID_PRODUCER_ID_MAPPING error, the server will not accept an EndTxnRequest, so skip
-        // directly to InitProducerId. Otherwise, we must first abort the transaction, because the producer will be
-        // fenced if we directly call InitProducerId.
-        if (!(lastError instanceof InvalidPidMappingException)) {
-            EndTxnRequest.Builder builder = new EndTxnRequest.Builder(
-                    new EndTxnRequestData()
-                            .setTransactionalId(transactionalId)
-                            .setProducerId(producerIdAndEpoch.producerId)
-                            .setProducerEpoch(producerIdAndEpoch.epoch)
-                            .setCommitted(transactionResult.id),
-                    isTransactionV2Enabled
-            );
+        // We must first abort the transaction, because the producer will be
```

Comment:
Is the comment correct?  Would we get fenced or just get concurrent transactions error?

Reply:
I had this thought too. I think we would get the concurrent transactions error first as we bump the epoch. Then we would need to call init producer ID again. So I think we do need to make the two calls, but the exact wording here isn't quite right.

Reply:
Oh yes, this was the existing comment and I just cut it down, was actually discussing with Calvin that the wording was right! Thanks for pointing it out, will change it!